Sources/MongoSwift/MongoConnectionString.swift

```diff
@@ -156,8 +181,109 @@ public struct MongoConnectionString: Codable, LosslessStringConvertible {
                     message: "A port cannot be specified in a mongodb+srv connection string")
             }
         }
+        let authAndOptions = identifiersAndOptions.count == 2 ?
+            identifiersAndOptions[1].components(separatedBy: "?") : nil
+        guard authAndOptions?.count ?? 0 <= 2 else {
+            throw MongoError.InvalidArgumentError(
+                message: "Connection string contains an unexpected '?'"
+            )
+        }
+        var database = try MongoConnectionString.percentDecode(from: authAndOptions?[0])
+        if (database ?? "").isEmpty {
+            database = nil
+        }
+        if userInfo != nil {
+            guard userInfo?.count ?? 0 <= 2 else {
+                throw MongoError.InvalidArgumentError(
+                    message: "Expected only a single auth info delimiter ':'"
+                )
+            }
+            self.credential = MongoCredential(
+                username: try MongoConnectionString.percentDecode(from: userInfo?[0]),
+                password: userInfo?.count == 2 ? try MongoConnectionString.percentDecode(from: userInfo?[1]) : nil,
+                source: database
+            )
+        }
+        let optionString = authAndOptions?.count ?? 0 == 2 ? authAndOptions?[1] : nil
+        if let options = optionString?.components(separatedBy: "&") {
+            for option in options {
+                let keyValue = option.components(separatedBy: "=")
+                guard keyValue.count == 2 else {
+                    throw MongoError.InvalidArgumentError(
+                        message: "Invalid key value pair for option \(keyValue)"
+                    )
+                }
+                let key = keyValue[0]
+                let value = keyValue[1]
+                switch key {
+                // authentication options.
+                case "authSource":
```

Comment:
related to my comment above, perhaps you could break out parsing auth options into a separate helper method, and figure out a way to consolidate the repeated check for whether `self.credential != nil`. 

I think what you're doing here to avoid multiple linear scans through the array is a good idea, so maybe you could continue supporting that by e.g. parsing out options into a `[String: String]` that you pass around into helpers that parse specific options?

Reply:
authOptions is just a temporary variable but is accessible if needed 
options will not be applied and verified in separate functions